### ISSUE TITLE
Training ResNet50 accuracy degradation on CIFAR10 using GACT

### ISSUE THREAD

**Issue by Zlzzzupup** [LiuXiaoxuanPKU/GACT-ICML#4]
I saw your work at ICML2022 and was very interested in it. When I tried to test it at resnet50, I found that there was a significant drop in training accuracy after training the same epoch using GACT. I don't know if there is something wrong with the GACT code I wrote, I have posted the code and results below. When I train 10 epochs by default without GACT, the accuracy can reach 70%, while with GACT it can only reach 55%.
```
def main():
    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    start_epoch = 0

    # Data
    print('==> Preparing data..')
    transform_train = transforms.Compose([
        transforms.RandomCrop(32, padding=4),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
    ])

    transform_test = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
    ])

    trainset = torchvision.datasets.CIFAR10(
        root='./data', train=True, download=True, transform=transform_train)
    trainloader = torch.utils.data.DataLoader(
        trainset, batch_size=256, shuffle=True, num_workers=2)

    testset = torchvision.datasets.CIFAR10(
        root='./data', train=False, download=True, transform=transform_test)
    testloader = torch.utils.data.DataLoader(
        testset, batch_size=100, shuffle=False, num_workers=2)

    classes = ('plane', 'car', 'bird', 'cat', 'deer',
            'dog', 'frog', 'horse', 'ship', 'truck')


    # GACT
    print('==> Building model..')
    net = ResNet50()
    net = net.to(device)

    gact.set_optimization_level('L1')
    controller = Controller(net)
    controller.install_hook()

    # optimizer
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(net.parameters(), lr=0.001,
                      momentum=0.9, weight_decay=5e-4)
    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=200)

    # Training
    for epoch in range(start_epoch, start_epoch + 10):
        print('\nEpoch: %d' % epoch)
        train_loss = 0
        correct = 0
        total = 0
        net.train()

        for batch_idx, (inputs, targets) in enumerate(trainloader):
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = net(inputs)
            loss = criterion(outputs, targets)

            def backprop():
                net.train()
                optimizer.zero_grad()
                outputs = net(inputs)
                loss = criterion(outputs, targets)
                loss.backward()

            optimizer.zero_grad()
            loss.backward()
            optimizer.step() # update the parameters
            controller.iterate(backprop) # tell gact how to perform forward/backward

            train_loss += loss.item()
            _, predicted = outputs.max(1)
            total += targets.size(0)
            correct += predicted.eq(targets).sum().item()

            progress_bar(batch_idx, len(trainloader), 'Loss: %.3f | Acc: %.3f%% (%d/%d)'
                     % (train_loss/(batch_idx+1), 100.*correct/total, correct, total))

        scheduler.step()
    controller.uninstall_hook()
```
```
==> Preparing data..
Files already downloaded and verified
Files already downloaded and verified
==> Building model..

Epoch: 0
 [=========================== 196/196 ============================>]  Step: 111ms | Tot: 1m5s | Loss: 2.059 | Acc: 22.308% (11154/50000)                                                                                              

Epoch: 1
 [=========================== 196/196 ============================>]  Step: 111ms | Tot: 1m5s | Loss: 1.762 | Acc: 33.416% (16708/50000)                                                                                              

Epoch: 2
 [=========================== 196/196 ============================>]  Step: 110ms | Tot: 1m5s | Loss: 1.628 | Acc: 39.020% (19510/50000)                                                                                              

Epoch: 3
 [=========================== 196/196 ============================>]  Step: 111ms | Tot: 1m5s | Loss: 1.520 | Acc: 43.562% (21781/50000)                                                                                              

Epoch: 4
 [=========================== 196/196 ============================>]  Step: 111ms | Tot: 1m5s | Loss: 1.436 | Acc: 47.374% (23687/50000)                                                                                              

Epoch: 5
 [=========================== 196/196 ============================>]  Step: 110ms | Tot: 1m5s | Loss: 1.376 | Acc: 49.914% (24957/50000)                                                                                              

Epoch: 6
 [=========================== 196/196 ============================>]  Step: 110ms | Tot: 1m5s | Loss: 1.327 | Acc: 51.632% (25816/50000)                                                                                              

Epoch: 7
 [=========================== 196/196 ============================>]  Step: 110ms | Tot: 1m5s | Loss: 1.295 | Acc: 53.016% (26508/50000)                                                                                              

Epoch: 8
 [=========================== 196/196 ============================>]  Step: 110ms | Tot: 1m5s | Loss: 1.272 | Acc: 54.066% (27033/50000)                                                                                              

Epoch: 9
 [=========================== 196/196 ============================>]  Step: 111ms | Tot: 1m5s | Loss: 1.251 | Acc: 55.280% (27640/50000) 
```

Looking forward to your reply.

**Comment by Zlzzzupup** [comment]
I read in the paper that the accuracy drop is <0.5%, is this achieved by training more epochs? I.e. it is impossible to achieve roughly the same accuracy result with the same epoch.
> We also vary the average bits used by the adaptive algorithm to explore the memory accuracy trade-off . On both tasks, GACT L1 achieves comparable (< 0:5% accuracy drop) or even better results than `the` full precision training, while reducing activation memory by up to 7.44 .

**Comment by LiuXiaoxuanPKU** [comment]
Thanks for the attention! optimization level 'L1' is the fix-4 bit baseline,  which will quantize all layers into fixed 4 bit.

To fix the problem, you can 

1. try to set the optimization level to 'L2', it will give different bits to different layers based on their sensitivity

or

2. change the condition of this [line](https://github.com/LiuXiaoxuanPKU/GACT-ICML/blob/5ddd88357bcf7ed8341a9b6a7be4a1ac4cdf88a9/gact/gact/quantizer.py#L64) from

`if ((len(input_tensor.shape) != 2)
            and (len(input_tensor.shape) != 3)
            and (len(input_tensor.shape) != 4)
            ):`
to
`if ((len(input_tensor.shape) != 3)
            and (len(input_tensor.shape) != 4)
            ):`
It will quantize all 3D and 4D tensors to 4bit. Based on our experience, for ResNet and GNNs, we don't want to quantize 2D tensors because they will affect accuracy greatly.

Try 1 or 2, either should work, please let me know if you still meet the problem, thanks!

**Comment by Zlzzzupup** [comment]
I am glad to receive your reply, I tried to set the optimization level to 'L2' and the problem of accuracy degradation was solved very well.

I still have some questions when I read the paper. About setting the optimization level of Resnet50 to 'L2', the training time for 10 epochs on CIFAR10 dataset is about 1000s, while the normal training without GACT is about 500s. Does this lead to a further increase in training time when using GACT in some larger scale model training? I don't seem to see any data in the paper about training time or throughput without GACT.

Looking forward to your reply.

**Comment by LiuXiaoxuanPKU** [comment]
Thanks for asking, it's a very good question. GACT is slow for two reasons:
(1) run extra iterations to get the sensitivity of different layers
(2) overhead introduced by quantization

you can observe (1) by seeing the output of running GACT, you will see a line similar to
`run extract XXX iterations`
the overhead of (2) is the overhead evaluated in the paper Figure 4.

I want to point out:
1. For (1), you can control the update frequency [here](https://github.com/LiuXiaoxuanPKU/GACT-ICML/blob/5ddd88357bcf7ed8341a9b6a7be4a1ac4cdf88a9/gact/gact/conf.py#L59). For example, changing it from 1000 to 10000, it will run extra XXX iterations every 10000 iterations instead of 1000.
3. For (2), we evaluated on T4, I expect more overhead on V100/A100, which is better at computation than memory access.

But in conclusion, you are right, GACT will introduce overhead (sometimes it might be big). You could play with the iteration parameter and hardware setting. Hope the answer helps!

Please let me know if you have more questions.

**Comment by Zlzzzupup** [comment]
Thanks for your replying.
Yes, I've noticed that 

> run extract XXX iterations

take up quite a bit of time in training, and adjusting the update frequency might be a good way to save time.
Finally, thank you very much for your answers, which gave me a deeper understanding of the work of GACT.
